### PR TITLE
WIP: first stab at macro by example

### PR DESCRIPTION
**Pull request by matklad**
The code is pretty ugly, but I am quite happy with the general organization that emerges. 

The center is a `tt::TokenTree` type, which is our version of proc_macro::TokenTree. It lives in a separate crate (`ra_macros`) which is independent of `hir` or `ra_syntax`, and which is pretty important. 

All macro interfaces should be formulated in terms of this `tt`. 

Alongside `tt`, lives a `mbe` (**m**acro **b**y **e**xample) module, which defines the AST for macro_rules. Ideally, this should live in yet another separate crate and depend on `ra_syntax`, to allow for `$e:expr` matchers. 

So, the workflow is this:

`ast::MacroCall` for `macro_rules` definition (`macro_rues impl_froms! {...}`) is parsed into `tt::TokenStream`. This `tt` is parsed into a mbe defenition: `mbe::Rules`

`ast::MacroCall` for macro invocation (`impl_froms!(Foo: Bar, Baz)`) is parsed into `tt::TokenStream` as well. Then, an mbe defenition is used to transform it into another `tt::TokenStream`, from which a `SyntaxNode` can be reconstruted.

### CODE REVIEW

File: crates/ra_macros/src/mbe_expander.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+            };
+        }
+        match b {
+            Binding::Simple(it) => Some(it),
+            Binding::Nested(_) => None,
+        }
+    }
+    fn push_nested(&mut self, nested: Bindings) -> Option<()> {
+        for (key, value) in nested.inner {
+            if !self.inner.contains_key(&key) {
```
Comment (kjeremy) [inline crates/ra_macros/src/mbe_expander.rs]: Is there an `entry` api we can use instead?

File: crates/ra_hir/src/macros.rs [hunk 2]
```diff
@@ -195,3 +197,102 @@ pub(crate) fn expand_macro_invocation(
...
+    let res = tt::Subtree {
+        delimiter,
+        token_trees,
+    };
+    Some(res)
+}
+
+#[test]
+fn test_convert_tt() {
+    let macro_defenition = r#"
```
Comment (kjeremy) [inline crates/ra_hir/src/macros.rs]: defenition -> definition in a bunch of places

File: crates/ra_macros/src/mbe_expander.rs (hunk 1 above)
Comment (matklad) [reply crates/ra_macros/src/mbe_expander.rs]: probably, though we'd still need that match agains `Nested`...

File: crates/ra_macros/src/mbe_parser.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
+use crate::{tt, mbe};
+use crate::tt_cursor::TtCursor;
+
+/// This module parses a raw `tt::TokenStream` into macro-by-example token
+/// stream. This is a *mostly* identify function, expect for handling of
+/// `$var:tt_kind` and `$(repeat),*` constructs.
```
Comment (kjeremy) [inline crates/ra_macros/src/mbe_parser.rs]: Good place for module docs

**Comment by DJMcNab** [comment]
It doesn't look like we handle desugaring doc comments into `#$(!)?[doc="..."]` yet, unless I'm mistaken. This is a probably another point for making doc comments more specialised.

File: crates/ra_mbe/src/syntax_bridge.rs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    let last_child = tt.last_child()?;
+    let delimiter = match (first_child.kind(), last_child.kind()) {
+        (L_PAREN, R_PAREN) => tt::Delimiter::Parenthesis,
+        (L_CURLY, R_CURLY) => tt::Delimiter::Brace,
+        (L_BRACK, R_BRACK) => tt::Delimiter::Bracket,
+        _ => return None,
+    };
+    let mut token_trees = Vec::new();
+    for child in tt.children().skip(1) {
+        if child == first_child || child == last_child || child.kind().is_trivia() {
```
Comment (bjorn3) [inline crates/ra_mbe/src/syntax_bridge.rs]: Doesn't this skip the inner `()` in the following code too?

```
( abc () )
```

Comment (matklad) [reply crates/ra_mbe/src/syntax_bridge.rs]: I think it should be handled by `let child: tt::TokenTree = if child.kind() == TOKEN_TREE`. But yeah, in general this all glosses over a **ton** of stuff. In this PR, I want to focus on creating a syntax -> tt -> mbe -> tt -> syntax interface. The `tt -> mbe` and `mbe -> tt` arrows could become `good-first-issue`s :) 

Here's a good test suite: https://github.com/intellij-rust/intellij-rust/blob/master/src/test/kotlin/org/rust/lang/core/macros/RsMacroExpansionTest.kt#L10-L680

**Comment by matklad** [comment]
Let's merge this!

It doesn't do anything useful yet, but the API sounds right!

There's to directions we should go from here:

* implement a more complete subset of macro_rules
* plug this into ra_hir

I'm going to concentrate on the second bit, so if anyone feels ambitions enough to dive into macro rules implementation, feel free to do it!

There's [a comment](https://github.com/rust-analyzer/rust-analyzer/blob/0c04e55ca23790d853986544f84fe5e53c61d3d6/crates/ra_mbe/src/lib.rs#L112-L117) explaining where to get a test suite from. It would also be cool to add a better error handling (currently we return Options everywhere, but we should use results) and the ability to match non-terminals like `$e:expr` (this is tricky, I think for starters we should just use string parsing for this, and adapt parser to token trees later). 

I would also love to point out [this comment](https://github.com/rust-analyzer/rust-analyzer/blob/0c04e55ca23790d853986544f84fe5e53c61d3d6/crates/ra_mbe/src/mbe_expander.rs#L19-L63) which explains one of the most interesting bits of the expansion process!

**Comment by matklad** [comment]
Can someone give a PR approval? :) Looks like it is no longer possible to self-approve the PR

**Comment by matklad** [comment]
bors try

**Comment by matklad** [comment]
merging manually after try

**Pull request merged**